src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Extensions.Caching.Memory
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce CacheEntry size by replacing many enum & boolean fields with one of a size of Int32
+        [StructLayout(LayoutKind.Explicit)]
+        private struct CacheEntryState
+        {
+            [FieldOffset(0)]
+            private byte _flags;
+            [FieldOffset(1)]
+            private byte _evictionReason;
+            [FieldOffset(2)]
+            private byte _priority;
+            [FieldOffset(3)]
+            private byte _reserved; // for future use
+
+            internal CacheEntryState(CacheItemPriority priority) : this() => _priority = (byte)priority;
+
+            internal bool IsDisposed
+            {
+                get => ((Flags)_flags & Flags.IsDisposed) != 0;
+                set => SetFlag(Flags.IsDisposed, value);
+            }
+
+            internal bool IsExpired
+            {
+                get => ((Flags)_flags & Flags.IsExpired) != 0;
+                set => SetFlag(Flags.IsExpired, value);
+            }
+
+            internal bool IsValueSet
+            {
+                get => ((Flags)_flags & Flags.IsValueSet) != 0;
+                set => SetFlag(Flags.IsValueSet, value);
+            }
+
+            internal EvictionReason EvictionReason
+            {
+                get => (EvictionReason)_evictionReason;
+                set => _evictionReason = (byte)value;
+            }
+
+            internal CacheItemPriority Priority
+            {
+                get => (CacheItemPriority)_priority;
+                set => _priority = (byte)value;
+            }
+
+            private void SetFlag(Flags option, bool value) => _flags = (byte)(value ? (_flags | (byte)option) : (_flags & ~(byte)option));
```

Comment:
It's atomic (https://github.com/dotnet/runtime/pull/45962#issuecomment-743388974), but what about the visibility of the change? I think it should be volatile read and writes. Also for the properties here.

Reply:
the `IsValueSet` and `IsDisposed` flags are set (and read) by a single thread that adds the entry to the cache, the `IsExpired` can, in theory, be set by two threads (one that is performing `cache.TryGetValue` and another that performs the background check for expired items) but even if the change is not visible to the other thread, it will just compare the dates and set `IsExpired` to true again (the `CheckExpired` logic)

Reply:
> `IsExpired` can, in theory, be set by two threads (one that is performing `cache.TryGetValue` and another that performs the background check for expired items) but even if the change is not visible to the other thread

The "double write" is not the problem. I see the problem when it comes to reading the value, as the fresh value may not be visible then.

Let's assume Thread A on CPU 1 [sets expired](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L163-L169) and the [bits of _flags are set](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs#L35).
Thread B on CPU 2 [checks for expiration](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L158-L159) which [reads the corresponding bit of _flags](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs#L34).

What guarantees that Thread B sees the freshest value of `_flags`? 
It could happen that the change in _flags gets trapped in CPU 1's cache, so that CPU 2 / Thread B see a stale value.

I believe especially on ARM with the weak memory model this could be a concern.
Yeah, `volatile` has a perf hit on ARM, but it's needed for correctness?
Instead of the `volatile`-field I'd use `Volatile.Read` and `Volatile.Write` (to make it more obvious what happens).
Maybe there's a better way this can be handled -- at the moment I have no (other) idea (except Interlocked, but that's even heavier, also for x86/x64).

[ScanForExpiredItems](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L311) check is concurrent, and this won't change with https://github.com/dotnet/runtime/pull/45842 [this call](https://github.com/dotnet/runtime/blob/14b3513cbe00e8306d305f7bcef6a5ae8d58acea/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L288).

Reply:
> What guarantees that Thread B sees the freshest value of _flags?

There is no such guarantee.

In this case, thread `B` might get `false` from `_state.IsExpired` so it's going to perform the full check:

https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L160

and find out that it's expired anyway. Moreover, as soon as cache entry is set to expired by any thread it also gets removed from the cache a moment later. So we are talking about an extra very rare check

```ini
// thread A                 Thread B
------------------------|---------------------------
entry.SetExpired()      |   entry.IsExpired 
entries.Remove(entry)   |   entry.CheckExpired
                        |   entries.Remove(entry)
```

Reply:
So in other words: `_state.IsExpired` is quick-check-hint to avoid the full check. In the case of a race, we fall back to the full check.
Is this correct? If so, then there's no problem.

Reply:
@gfoidl exactly! thank you for a great discussion

Reply:
Thank you too! Always nice to discuss about memory models.